planner/core/preprocess.go

```diff
@@ -1420,16 +1418,24 @@ func (p *preprocessor) checkFuncCastExpr(node *ast.FuncCastExpr) {
 	}
 }
 
-// handleAsOf tries to validate the timestamp.
-// If it is not nil, timestamp is used to get the history infoschema from the infocache.
-func (p *preprocessor) handleAsOf(node *ast.AsOfClause) {
-	readTS := p.ctx.GetSessionVars().TxnReadTS.PeakTxnReadTS()
-	if readTS > 0 && node != nil {
-		p.err = ErrAsOf.FastGenWithCause("can't use select as of while already set transaction as of")
-		return
+// handleAsOfAndReadTS tries to handle as of closure, or possibly read_ts.
+// If read_ts is not nil, it will be consumed.
+// If as of is not nil, timestamp is used to get the history infoschema from the infocache.
+func (p *preprocessor) handleAsOfAndReadTS(node *ast.AsOfClause) {
```

Comment:
I think `handleAsOfAndReadTS` is much complex now. We need to add example to explain it and add unit test to ensured logic.